### PR TITLE
[FABN-1564] SubjectDN extensibility

### PR DESCRIPTION
**Pull request by davidkhala**
In multiple places we have only common name as input for building X509
Subject.
We should allow directly full input of subject

Signed-off-by: david <david-khala@hotmail.com>

### CODE REVIEW

File: fabric-ca-client/test/FabricCAServices.js [hunk 1]
```diff
@@ -249,7 +249,7 @@ describe('FabricCAServices', () => {
 
 		it('should reject on enroll failure', async () => {
 
-			getSubjectCommonNameStub.returns('mr_penguin');
+			getSubjectStub.returns('mr_penguin');
```
Comment (bestbeforetoday) [inline fabric-ca-client/test/FabricCAServices.js:252]: The return value should probably be `/CN=mr_penguin` to make it a valid subject string. I know it doesn't break the code today but might confuse somebody in the future if they change the code do something with the subject and all these tests start failing.

File: fabric-ca-client/lib/helper.js [hunk 2]
```diff
@@ -28,23 +28,27 @@ function checkRegistrar(registrar) {
...
 // fails to parse certs that includes ECDSA keys.
-function getSubjectCommonName(pem) {
+const getSubject = (pem) => {
 	const hex = jsrsasign.pemtohex(pem);
 	const d = ASN1HEX.getTLVbyList(hex, 0, [0, 5]);
-	const subject = X509.hex2dn(d); // format: '/C=US/ST=California/L=San Francisco/CN=Admin@org1.example.com/emailAddress=admin@org1.example.com'
+	return X509.hex2dn(d); // format: '/C=US/ST=California/L=San Francisco/CN=Admin@org1.example.com/emailAddress=admin@org1.example.com'
+};
+
+const getSubjectCommonName = (pem) => {
```
Comment (bestbeforetoday) [inline fabric-ca-client/lib/helper.js:43]: I don't _think_ this is used anywhere. if not, might as well remove it. Just be careful to make sure the corresponding unit tests still drive `getSubject()` instead where appropriate.

**Review by bestbeforetoday — CHANGES_REQUESTED** [review]
Generally looks good to me. Just some minor tweaks

File: fabric-ca-client/lib/FabricCAServices.js [hunk 3]
```diff
@@ -136,16 +136,18 @@ const FabricCAServices = class extends BaseClient {
...
 	 * @property {string} enrollmentSecret - The secret associated with the enrollment ID
-	 * @property {string} profile - The profile name.  Specify the 'tls' profile for a TLS certificate;
+	 * @property {string} [profile] - The profile name.  Specify the 'tls' profile for a TLS certificate;
 	 *                   otherwise, an enrollment certificate is issued.
-	 * @property {string} csr - Optional. PEM-encoded PKCS#10 Certificate Signing Request. The message sent from client side to
+	 * @property {string} [csr] - Optional. PEM-encoded PKCS#10 Certificate Signing Request. The message sent from client side to
 	 *                   Fabric-ca for the digital identity certificate.
-	 * @property {AttributeRequest[]} attr_reqs - An array of {@link AttributeRequest}
+	 * @property {AttributeRequest[]} [attr_reqs] - An array of {@link AttributeRequest}
+	 * @property {string} subject - Optional. The X509 Subject to use in generating CSR inline
```
[the remark this answers is not in the document]
Comment (davidkhala) [reply fabric-ca-client/lib/FabricCAServices.js:144]: Definitely

**Comment by davidkhala** [comment]
> Generally looks good to me. Just some minor tweaks

Thanks again. Agree all all feedback. WIP

**Review by bestbeforetoday — APPROVED** [review]

**Pull request merged**